### PR TITLE
Read header and footer if available

### PR DESCRIPTION
**Pull request by sebastian-alfers**
This commit fixes an issue, that header-values can not be restored correctly after an event was avro-serialized to disk.

The problem was, that deserializing the event from disk, the body of the event contained the binary and the header.

See: https://issues.apache.org/jira/browse/FLUME-2942

Discussion welcome!

### CODE REVIEW

File: flume-ng-core/src/main/java/org/apache/flume/serialization/AbstractAvroEventDeserializer.java [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+
+/**
+ * A deserializer that parses Avro container files, generating one Flume event
+ * per record in the Avro file, and storing binary avro-encoded records in
+ * the Flume event body.
+ */
+public abstract class AbstractAvroEventDeserializer implements EventDeserializer {
+
+  protected AvroSchemaType schemaType;
+  protected ResettableInputStream ris;
```
Comment (adenes) [inline flume-ng-core/src/main/java/org/apache/flume/serialization/AbstractAvroEventDeserializer.java]: It would be better to change this to `private final`: just add it to the constructor and replace the `this.ris = ris` with a `super(ris)` call in the subclasses.
There is no need for the subclasses to access this and pushing the responsibility of setting it (and also giving the possibility to modify it) to the subclasses would make it very error-prone.

File: flume-ng-core/src/main/java/org/apache/flume/serialization/AbstractAvroEventDeserializer.java [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+  }
+
+  public static final String CONFIG_SCHEMA_TYPE_KEY = "schemaType";
+  public static final String AVRO_SCHEMA_HEADER_HASH
+      = "flume.avro.schema.hash";
+  public static final String AVRO_SCHEMA_HEADER_LITERAL
+      = "flume.avro.schema.literal";
+
+
+  protected void initialize(ResettableInputStream ris)
```
Comment (adenes) [inline flume-ng-core/src/main/java/org/apache/flume/serialization/AbstractAvroEventDeserializer.java]: The parameter is not needed if `AbstractAvroEventDeserializer.ris` is changed to `private final`.

File: flume-ng-core/src/main/java/org/apache/flume/serialization/EventDeserializerType.java [hunk 3]
```diff
@@ -26,6 +26,7 @@
 public enum EventDeserializerType {
   LINE(LineDeserializer.Builder.class),
   AVRO(AvroEventDeserializer.Builder.class),
+  FLUME(FlumeEventAvroEventDeserializer.Builder.class),
```
Comment (adenes) [inline flume-ng-core/src/main/java/org/apache/flume/serialization/EventDeserializerType.java]: I think `FLUME_AVRO` would be a more expressive name for this.
Please add this to the user guide as well. Thanks.

File: flume-ng-core/src/main/java/org/apache/flume/serialization/FlumeEventAvroEventDeserializer.java [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+ */
+public class FlumeEventAvroEventDeserializer extends AbstractAvroEventDeserializer {
+
+  private static final Logger logger =
+          LoggerFactory.getLogger(FlumeEventAvroEventDeserializer.class);
+
+  private FlumeEventAvroEventDeserializer(Context context, ResettableInputStream ris) {
+    this.ris = ris;
+
+    schemaType = AvroSchemaType.valueOf(
```
Comment (adenes) [inline flume-ng-core/src/main/java/org/apache/flume/serialization/FlumeEventAvroEventDeserializer.java]: This (line 48-57) is a copy-paste from `AvroEventDeserializer`. Could you please simplify it by pushing it (from here and from the `AvroEventDeserializer`) up to the `AbstractAvroEventDeserializer`? 
If the logic is moved to the parent class then its `schemaType` field can be changed to `private final`.

File: flume-ng-core/src/main/java/org/apache/flume/serialization/AbstractAvroEventDeserializer.java [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+          throws IOException, NoSuchAlgorithmException {
+    SeekableResettableInputBridge in = new SeekableResettableInputBridge(ris);
+    long pos = in.tell();
+    in.seek(0L);
+    fileReader = new DataFileReader<GenericRecord>(in,
```
Comment (adenes) [inline flume-ng-core/src/main/java/org/apache/flume/serialization/AbstractAvroEventDeserializer.java]: nit: explicit type argument can be replaced with `<>`

**Review by adenes — COMMENTED** [review]
@sebastian-alfers, thank you for the pull request, I had some comments on it. Let me know if you have any question regarding to them.

**Comment by sebastian-alfers** [comment]
@adenes Thanks fore your comments, I updated the pr

**Comment by sebastian-alfers** [comment]
@adenes any comments?

**Comment by sebastian-alfers** [comment]
@adenes ping

**Comment by sebastian-alfers** [comment]
@laxman-ch @adenes @simonati anyone likes to push this PR forward? Comments welcome

**Comment by simonati** [comment]
Thanks for the heads up! Let me take a look today.

**Comment by simonati** [comment]
Hi @sebastian-alfers,

This change is quite big and hard to follow. Please give me more time to finish the review or help any other reviewer by making consumption of your changes easier by leaving some more hints about the ideas behind your changes.

**Comment by sebastian-alfers** [comment]
@simonati Thanks for you comment. Actually, I wrote down all my ideas in the corresponding jira ticket. If you have any particular question feel free to point them out

**Comment by sebastian-alfers** [comment]
Push

**Comment by asfgit** [comment]
Can one of the admins verify this patch?
